Make new apps "deployable to Heroku" by default

sebarmano opened the pull request:
This change makes new suspender apps "deployable to Heroku" by default,
moving the Heroku setup in the local machine to be run by default (i.e.
with the --heroku flag off). All the setup that requires to hit the
Heroku API to create the staging and production apps is kept to be run
only with the flag on.

tute commented:
Thank you! Linking to related issues:

* https://github.com/thoughtbot/suspenders/issues/735
* https://github.com/thoughtbot/suspenders/pull/604

tute commented:
While we are editing, can we improve the changed lines to comply with the style guide, and please Hound CI? Thanks!

tute commented on spec/features/new_project_spec.rb:
```diff
@@ -223,6 +223,43 @@
+
+  it "adds app.json file" do
+    expect(File).to exist("#{project_path}/app.json")
+  end
```
This spec is a subset of the next one (if this fails, the next one sure will). We can drop it.

sebarmano replied:
Yes, I'll remove this test. Also, I think that the app.json test should be inside the "creates the local heroku setup" block, as it is part of it. What do you think?

tute replied:
> Also, I think that the app.json test should be inside the "creates the local heroku setup" block, as it is part of it. What do you think?

That's true, but I like having smaller specs with more specific descriptions. I'd leave them separate (even though the project does not follow this rule in the different spec files).

tute commented on spec/features/new_project_spec.rb:
```diff
@@ -223,6 +223,43 @@
+
+    expect(bin_deploy).to include("heroku run rake db:migrate --exit-code")
+    expect(File.stat(bin_deploy_path)).to be_executable
+
```
Let's add an `it` section:

> it "creates a circle.yml file with staging deployment commands" do

tute replied:
> it "creates a circle.yml file with staging deployment commands" do

Now that I think of this, it will fail if there is no Heroku app set up. Circle CI configuration should only happen with `--heroku true` flag.

sebarmano replied:
I missed this comment. I'll move the Circle CI confuguration back for it to be under the --heroku flag.

tute commented:
Left a few comments and questions. This is great work, thank you @sebarmano!

tute commented:
This is already a step forward. Question: do you think it's worth it to extract a shell script that performs the stateful operations?

* Create Heroku apps, pipelines, and remotes
* Configure CI autodeploys
* Set static files, rails secret, and application host constants

This would require some refactoring in suspenders. We could work on it as part of a different PR, too.

tute commented:
Great, thanks! Let me know when you consider this ready for another review.

sebarmano commented:
> Let me know when you consider this ready for another review.

I think it's ready for review @tute. I fixed the Hound issues and deleted the redundant app.js test. Please, let me know if you think of anything else that can be improved regarding this PR.

tute commented:
> I think it's ready for review @tute. I fixed the Hound issues and deleted the redundant app.js test. Please, let me know if you think of anything else that can be improved regarding this PR.
>  Show all checks

What do you think of dividing the spec with more descriptive sections? https://github.com/thoughtbot/suspenders/pull/756#discussion_r61895648

tute commented:
Do you think we should solve https://github.com/thoughtbot/suspenders/pull/756#issuecomment-216561336 as part of another PR?

sebarmano commented:
> What do you think of dividing the spec with more descriptive sections?
I think it's a good idea. Right now we are "suspending" only one app to run all the tests in the block. Splitting the tests will add the time of creating a new app for each one of them (which I don't think is an issue at this point given the amount of tests). 
Also, right now there are three tests inside the block and one outside (the app.js one). I think splitting the tests will improve consistency in the feature spec file.

> Do you think we should solve #756 (comment) as part of another PR?
As I understand, the idea is to be able to have fully functional apps without the need of creating the staging and production Heroku counterparts. I think that if we keep the circle.yml file by default and someone intends to use Circle CI it just for their local app, it won't work (please correct me if I'm wrong on this). So I think the best call is to keep that file to be created only under when the --heroku flag is on. Also, I think that the change is part of this same PR as otherwise we would be merging something that is not 100% functional for some use cases.

sebarmano commented:
Ok, I think that the PR is ready for another review. I moved back the Circle CI configuration based on [this comment](https://github.com/thoughtbot/suspenders/pull/756#discussion_r61896289). Also, I've divided the specs in new_project feature spec file to be more descriptive.

tute commented:
> Splitting the tests will add the time of creating a new app for each one of them

Because we use `before(:all)` instead of `before(:each)`, we recreate a suspenders app once regardless of the number of examples in the spec file, so there should be no performance penalty.

> I think that we can work on that in another PR, as it is related to this same issue but is not exactly the same thing. This PR is solving a problem related to not having the Heroku setup for local apps. The comment you mention seems more like a new feature of suspenders, right?

Makes sense to me. 👍

tute commented on spec/features/new_project_spec.rb:
```diff
@@ -223,6 +223,33 @@
+    expect(File.stat(bin_setup_path)).to be_executable
+  end
+
+  it "creates heroku pipelines config file" do
```
This description should be something like "creates deploy script".

tute commented on lib/suspenders/generators/app_generator.rb:
```diff
@@ -165,18 +166,22 @@ def setup_git
+      say "Creating local Heroku setup"
+      build :provide_review_apps_setup_script
+      build :create_heroku_pipelines_config_file
+      build :provide_deploy_script
```
While we are on this, should we make these names more homogeneous? They all create files, but two start with `provide` and one with `create`.

sebarmano replied:
Yes, makes sense. I'll change this and what you mention in the previous comment.

tute commented:
Looking great, left two minor comments. Thank you! 🎉

sebarmano commented:
I just pushed the last changes and I think it's again ready for another review. In addition to the changes you mentioned, I changed the name of the methods and tests related to app.js file to mention the "manifest". I think that name is more accurate and is the name Heroku uses. In this way, for example, if someone wants to know more about what this file does, it will be easier to google for it after the methods' names and their contents. Feel free to tell me if this doesn't make sense, I can rename back the methods.

tute commented on lib/suspenders/generators/app_generator.rb:
```diff
@@ -165,18 +166,22 @@ def setup_git
+      say "Creating local Heroku setup"
+      build :create_review_apps_setup_script
+      build :create_heroku_application_manifest_file
+      build :create_deploy_script
```
We need to move up here (and the related spec) the `set_up_heroku_specific_gems` step.

Thank you!

tute commented:
Looking great! Added a note on a missing step. Thanks!

sebarmano commented:
Ready for another review (thanks for taking the time to review this like a million times!). I moved the Heroku gems method to be run by default. 
The only thing that I'm not 100% sold with is the test. I left the test as it was before (just moved it to the right place), but it's testing just that the gem is in the Gemfile, not that is inside the production & staging block. I couldn't find a way of testing it in an easy way, but if you think of one I'm happy to add that too.

tute commented:
Hey, thank you for your patience with the iterative work and feedback! :)

You know what? If we always add that gem into the Gemfile, we can directly edit the `template/Gemfile.erb` file, and add the gem in the production environment, after `rack-timeout` so we keep the alphabetical order. We can then drop the code for setup heroku gems, but I'd leave the test where it is.

Thanks again for your work and patience! :)

sebarmano commented:
Sure! This is good because I learn how the codebase works with each step.
I just pushed another version for review. As we removed the method that was adding the `rails_stdout_logging` I was able to also remove a method from the Heroku adapter and its test that were there just because of that gem. This is just a heads up so you can tell me if I went too far removing that.

tute commented:
The spec would document why we have that line there, and in that sense it is useful. Anyway Rails 5 is just around the corner and we won't need that gem anymore, so this looks good. Will test soon-ish, and when I see everything works as we expect, I'll merge. Thank you! 🎉

sebarmano commented:
Just to clarify, I didn't remove the test in the new project feature spec referencing that gem, which is still there and testing it. 
In the heroku adapter there was a method to add that gem (and the corresponding spec in the heroku adapter tests). I removed that method (and test) which I think is not needed anymore because the gem is added via the Gemfile template. 

Also, just found a typo in that I'll fix now.

tute commented:
> Just to clarify, I didn't remove the test in the new project feature spec referencing that gem, which is still there and testing it. 
> In the heroku adapter there was a method to add that gem (and the corresponding spec in the heroku adapter tests). I removed that method (and test) which I think is not needed anymore because the gem is added via the Gemfile template.

That is perfect. 👏 👍

sebarmano commented:
Just fixed the typo in this last push, so it's basically the same as before.

tute commented:
Thanks so much! 🎉
